pkg/core/blockchain.go

```diff
@@ -1516,6 +1533,51 @@ func (bc *Blockchain) removeOldHeaderHashes(index uint32) time.Duration {
 	return dur
 }
 
+func (bc *Blockchain) removeUntraceableBlocks(newPeriod, tgtHeight uint32) time.Duration {
+	if newPeriod*bc.config.GarbageCollectionPeriod/headerBatchCount == tgtHeight/headerBatchCount { // current batch of header hashes is not stored yet.
+		tgtHeight = uint32(max(0, (int(tgtHeight)/headerBatchCount-1)*headerBatchCount))
+	}
+	if tgtHeight == 0 {
+		return 0
+	}
+	oldTgtHeight := atomic.LoadUint32(&bc.gcLastUntraceableBlockHeight)
```

Comment:
@roman-khimov this solution looks kludgy to me, but I can't offer anything better than binding old untraceable blocks removal target height to the blockchain. The problem is that we need to know the exact height of the previous run, and it's not possible to retrieve it from the old persisted height that is passed to GC routine.

Reply:
Why is it atomic then? `init()` is not running concurrently to `tryRunGC()`.